primefaces/src/main/java/org/primefaces/component/api/PrimeUIData.java

```diff
@@ -0,0 +1,521 @@
+        }
+        else if (phaseId == PhaseId.UPDATE_MODEL_VALUES) {
+            component.processUpdates(context);
+        }
+
+    }
+
+    @Override
+    protected void setRowIndexWithoutRowStatePreserved(int rowIndex) {
+        saveDescendantState();
+        setRowModel(rowIndex);
+        restoreDescendantState();
+    }
+
+    public void setRowModel(int rowIndex) {
+        //update rowIndex
+        getStateHelper().put(UIDataPatch.PropertyKeys.rowIndex, rowIndex);
+        getDataModel().setRowIndex(rowIndex);
+
+        //clear datamodel
+        if (rowIndex == -1) {
+            setDataModel(null);
+        }
+
+        //update var
+        String var = getVar();
+        if (var != null) {
+            String rowIndexVar = getRowIndexVar();
+            Map<String, Object> requestMap = getFacesContext().getExternalContext().getRequestMap();
+
+            if (rowIndex == -1) {
+                oldVar = requestMap.remove(var);
+
+                if (rowIndexVar != null) {
+                    requestMap.remove(rowIndexVar);
+                }
+            }
+            else if (isRowAvailable()) {
+                requestMap.put(var, getRowData());
+
+                if (rowIndexVar != null) {
+                    requestMap.put(rowIndexVar, rowIndex);
+                }
+            }
+            else {
+                requestMap.remove(var);
+
+                if (rowIndexVar != null) {
+                    requestMap.put(rowIndexVar, rowIndex);
+                }
+
+                if (oldVar != null) {
+                    requestMap.put(var, oldVar);
+                    oldVar = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void saveDescendantState() {
+        FacesContext context = getFacesContext();
+
+        if (getChildCount() > 0) {
+            for (int i = 0; i < getChildCount(); i++) {
+                UIComponent kid = getChildren().get(i);
+                saveDescendantState(kid, context);
+            }
+        }
+
+        if (getFacetCount() > 0) {
+            for (UIComponent facet : getFacets().values()) {
+                saveDescendantState(facet, context);
+            }
+        }
+    }
+
+    @Override
+    protected void restoreDescendantState() {
+        FacesContext context = getFacesContext();
+
+        if (getChildCount() > 0) {
+            for (int i = 0; i < getChildCount(); i++) {
+                UIComponent kid = getChildren().get(i);
+                restoreDescendantState(kid, context);
+            }
+        }
+
+        if (getFacetCount() > 0) {
+            for (UIComponent facet : getFacets().values()) {
+                restoreDescendantState(facet, context);
+            }
+        }
+    }
+
+    protected boolean shouldSkipChildren(FacesContext context) {
+        return false;
+    }
+
+    @Override
+    public boolean invokeOnComponent(FacesContext context, String clientId, ContextCallback callback)
```

Comment:
this could acutally inside UIDataPatch - or even removed if its finally fixed in all Mojarra versions, it implemented correctly in MyFaces